cpp/src_prims/random/make_blobs.cuh

```diff
@@ -26,46 +28,120 @@
 namespace MLCommon {
 namespace Random {
 
+namespace {
+
+// generate the labels first and shuffle them instead of shuffling the dataset
+template <typename IdxT>
+void generate_labels(IdxT* labels, IdxT n_rows, IdxT n_clusters, bool shuffle,
+                     cudaStream_t stream) {
+  // always keep 'a' to be coprime to n_rows
+  IdxT a = rand() % n_rows;
+  while (gcd(a, n_rows) != 1) a = (a + 1) % n_rows;
+  IdxT b = rand() % n_rows;
+  auto op = [=] __device__(IdxT * ptr, IdxT idx) {
+    if (shuffle) {
+      idx = IdxT((a * int64_t(idx)) + b) % n_rows;
+    }
+    *ptr = idx % n_clusters;
+  };
+  LinAlg::writeOnlyUnaryOp<IdxT, decltype(op), IdxT>(labels, n_rows, op,
+                                                     stream);
+}
+
+template <typename DataT, typename IdxT>
+DI void get_mu_sigma(DataT& mu, DataT& sigma, IdxT idx, const IdxT* labels,
+                     bool row_major, const DataT* centers,
+                     const DataT* cluster_std, DataT cluster_std_scalar,
+                     IdxT n_rows, IdxT n_cols, IdxT n_clusters) {
+  IdxT cid, fid;
+  if (row_major) {
+    cid = idx / n_cols;
+    fid = idx % n_cols;
+  } else {
+    cid = idx % n_rows;
+    fid = idx / n_rows;
+  }
+  IdxT center_id;
+  if (cid < n_rows) {
+    center_id = labels[cid];
+  } else {
+    center_id = 0;
+  }
+  if (row_major) {
+    center_id = center_id * n_cols + fid;
+  } else {
+    center_id += fid * n_clusters;
+  }
+  sigma = cluster_std == nullptr ? cluster_std_scalar : cluster_std[cid];
+  mu = centers[center_id];
+}
+
 template <typename DataT, typename IdxT>
-__global__ void gatherKernel(DataT* out, const DataT* in, const IdxT* perms,
-                             IdxT len) {
-  IdxT tid = blockIdx.x * blockDim.x + threadIdx.x;
-  if (tid < len) out[tid] = in[perms[tid]];
+void generate_data(DataT* out, const IdxT* labels, IdxT n_rows, IdxT n_cols,
+                   IdxT n_clusters, cudaStream_t stream, bool row_major,
+                   const DataT* centers, const DataT* cluster_std,
+                   const DataT cluster_std_scalar, Rng& rng) {
+  auto op = [=] __device__(DataT & val1, DataT & val2, IdxT idx1, IdxT idx2) {
+    DataT mu1, sigma1, mu2, sigma2;
+    get_mu_sigma(mu1, sigma1, idx1, labels, row_major, centers, cluster_std,
+                 cluster_std_scalar, n_rows, n_cols, n_clusters);
+    get_mu_sigma(mu2, sigma2, idx2, labels, row_major, centers, cluster_std,
+                 cluster_std_scalar, n_rows, n_cols, n_clusters);
+    constexpr auto twoPi = DataT(2.0) * DataT(3.141592654);
```

Comment:
I don't know how this would be feasible, but is it possible to reduce code duplicated from https://github.com/rapidsai/cuml/blob/branch-0.14/cpp/src_prims/random/rng.cuh#L172? I realise we want to calculate mu-sigma while we are in device context. What are your thoughts here?

Reply:
good point. Better to put this in a device method and use it everywhere else. Done.